presidio_evaluator/data_generator/presidio_data_generator.py

```diff
@@ -295,48 +292,119 @@ def name_gendered(row):
         return fake_data
 
 
-if __name__ == "__main__":
-    PresidioDataGenerator.seed(42)
-
-    template_file_path = Path(Path(__file__).parent, "raw_data", "templates.txt")
-
-    # Read FakeNameGenerator data
-    fake_data_df = pd.read_csv(
-        Path(Path(__file__).parent, "raw_data", "FakeNameGenerator.com_3000.csv")
-    )
-    # Convert column names to lowercase to match patterns
-    fake_data_df = PresidioDataGenerator.update_fake_name_generator_df(fake_data_df)
-
-    # Create a RecordsFaker (Faker object which prefers samples multiple objects from one record)
-    faker = RecordsFaker(records=fake_data_df, local="en_US")
-    faker.add_provider(IpAddressProvider)
-    faker.add_provider(NationalityProvider)
-    faker.add_provider(OrganizationProvider)
-    faker.add_provider(UsDriverLicenseProvider)
-    faker.add_provider(AgeProvider)
-    faker.add_provider(AddressProviderNew)  # More address formats than Faker
-    faker.add_provider(PhoneNumberProviderNew)  # More phone number formats than Faker
-
-    # Create Presidio Data Generator
-    data_generator = PresidioDataGenerator(custom_faker=faker, lower_case_ratio=0.05)
-    data_generator.add_provider_alias(provider_name="name", new_name="person")
-    data_generator.add_provider_alias(
-        provider_name="credit_card_number", new_name="credit_card"
-    )
-    data_generator.add_provider_alias(
-        provider_name="date_of_birth", new_name="birthday"
-    )
-
-    sentence_templates = PresidioDataGenerator.read_template_file(template_file_path)
-    fake_patterns = data_generator.generate_fake_data(
-        templates=sentence_templates, n_samples=10000
-    )
-
-    # save to json
-    output_file = Path(
-        Path(__file__).parent.parent.parent, "data", "presidio_data_generator_data.json"
-    )
+class PresidioFakeRecordGenerator:
```

Comment:
@omri374 @melmatlis
Easier to have a thread here to discuss the naming of this class and `PresidioDataGenerator` (and can have a follow-up call as @omri374 suggested)

My instinct is to rename both classes.

Wrt `PresidioDataGenerator` I find it surprising that it returns `Faker` entity types; as its name suggests, it'll return data in the "Presidio" form.
I also think the name could be made more concrete, as the term "Data" is very general, not giving a sense of what will be returned
I think the key features of this class, that could be captured in its name are:
- It is a wrapper around Faker, extending its functionality
- It returns `FakeSpan`s
- It works with `templates` - Critically so that it can maintain semantic dependency across sentences

I'd maybe even go with something as simple as `SentenceFaker`
Which captures the difference between this and `Faker` and when you would want to use one versus the other

Wrt `PresidioFakeRecordGenerator`, the key features of this class vs the other are:
- Uses `presidio` not `faker` entity-types
- Defaults to using all additional `presidio` entity providers and templates

So could be `PresidioSentenceFaker`
Distinguishing how `SentenceFaker` is in the `faker` domain and this class is the same idea but in the `presidio` domain

What are your thoughts?

Reply:
I suggest we have a call on this topic. 
Ideally I would like to design a single class which enables users to:
1. Use Faker as well as custom providers
2. Use default as well as user provided templates
3. Enable mappings between template <-> faker <-> presidio entity names

Reply:
Let's have a call. @Robbie-Palmer could you please email presidio@microsoft.com and we'll set up a call?

Reply:
@omri374 and @melmatlis to summarise what we discussed about this PR in the call
The current `PresidioDataGenerator` API looks like:
- Construct a `Faker` object and pass to constructor
- Choose locale and lower case ratio
- Generate faker spans (with faker entity types) from a list of templates
- Parse a single template string 
- `add_provider_alias`

The API in this PR looks like:
- Choose your
    - Locale
    - Lower case ratio
    - Entity providers
    - Sentence templates
    - Aliases (via a class field rather than via the constructor)
- Default to using all providers and templates within this repo
- Generate faker spans with presidio entity types

The fact that the `PresidioDataGenerator` takes a `Faker` object into its constructor and returns `Faker` entity types is why I think of it as being in the "Faker domain"
The new API in this PR doesn't expose that `Faker` is being used at all, keeping it in the "Presidio domain", which ties in with what @melmatlis said about wanting to expand to using other libraries with similar but extended functionality

So the differences between the two APIs at the moment are:
- The original requires construction of a `Faker` object by the caller
    - I don't know what use cases there may be for this, which is what I was referring to when saying perhaps researchers want this lower level API? But maybe they don't need it, if there are plans to abstract away the word faking interface to enable use of other libraries too
-  Returning faker entity types vs presidio entity types
- Original can parse a single template string, while new API cannot atm (can easily be added) 

So the questions I have are:
- Is there a use case where a user of this library wants to construct their own `Faker` instance and have `Faker` entity types returned?
    - If yes, should this be pulled into an explicitly `Faker` library oriented module, especially if there are plans to support alternative libraries? And maybe even pushed upstream to `Faker` itself in the long term?
    - If no, then are we happy to deprecate the old API and use the new one instead, therefore melding into one class?

Reply:
Hi @Robbie-Palmer. Thanks for the great summary. Initially there was an ambition to expose a `Faker` functionality to make it easy to use Faker objects, but I agree that this is not needed.

My suggestion is to proceed with the two classes approach (which is what this PR currently proposes, correct?), where one is more of the API and the other is wrapping Faker. The Faker wrapper isn't something that someone would typically use, and it further could be replaced if someone wants to use a different fake sentence generator.

If we go with two classes then with regards to naming the classes and the abstraction of Faker, we can consider calling the Faker wrapper class something with `Faker` in it. Maybe SpanFaker, TemplateToFakeDataGenerator or FakeSpanGenerator? (Although the span itself isn't fake, just the text... I'm not sure. Naming is always difficult :) )

I've left a few additional comments to still allow further customizability without having to inherit the class or change the code, but open to your thoughts. Thanks!

Reply:
I was trying to avoid doing a large refactor in case this became one of those PRs that lives too long and doesn't get merged 😅 
But I have went for it now
I've removed the overlap that was between the classes, renamed them and moved what remains of the old one into the `faker_extensions` package, consolidating it into a module with related classes

Reply:
Thank you! We didn't mean for this to be too big. Appreciate all the work!

Reply:
No problem! My main concern isn't the amount of work; it's more that the larger the changes, the more subjective it becomes. And the subjectivity of your team is more important than mine since long-term maintenance depends on the maintainers having a shared vocabulary / ubiquitous language
Plus changes I make could clash with internal work/communication I'm not aware of, so is easy to make broad sweeping changes that clash and then slow progress down

Also there's the worry that starting to refactor some things leads to a snowball effect
E.g. now in `PresidioSentenceFaker` I'm importing `FakerSpansResult as FakeSentenceResult` to avoid the "Faker domain" from leaking into this module
But looking at `FakerSpanResult` it could be renamed generally, as nothing about it is `Faker` specific other than it holding `List[FakerSpan]`
But looking at `FakerSpan` nothing about it is `Faker` specific
So both those classes could be renamed to be more generic, and the `data_objects` module that contains them should then be moved out of the `faker_extensions` package

There are also other examples like that, so the other main concern is finding a good cut-off point for this piece of work, and ensuring any further refactorings are done incrementally
Especially so that the bug fixes are delivered sooner rather than later to avoid other users being affected :)

Reply:
Completely agree Robbie with your last comment.
@omri374, I think you've mentioned you created a feature branch for all the refactoring we're doing on the code.
What if we change the target of the this PR to the feature branch, by this we will achieve 2 things
1. we won't break main 
2. It will enable us to build on top of Robbie's excellent modifications without the "stress" of prolonging the PR
What do you think?

Reply:
Yes, pushing everything into a feature branch would be a good idea. It would also allow me to test this locally without having to clone @Robbie-Palmer's fork :)

Reply:
Done. Changed the target branch to data-generator-2.1

Reply:
I also completely agree. I suggest to merge this into a feature branch, and plan any changes we see that are needed (such as the one you mentioned @Robbie-Palmer in the faker objects) prior to going and implementing them. If you agree, we'll continue the work on the feature branch also as there are other changes happening on the Evaluator side.

Reply:
Yes I'm happy with that :)

Reply:
Thanks :)

Reply:
BTW, this looks great. In my view could be merged to the main branch. We're putting it in a feature branch so that we could keep the consistency with the other parts that are changing (the evaluator part) @Robbie-Palmer @melmatlis @tranguyen221